lib/round-service.js

```
@@ -0,0 +1,185 @@
+
+  /**
+   * Stop the round service
+   */
+  stop () {
+    if (this.#checkRoundIntervalId) clearInterval(this.#checkRoundIntervalId)
+    console.log('Round service stopped')
+  }
+
+  /**
+   * Initialize the current round
+   */
+  async #initializeRound () {
+    const activeRound = await this.#getActiveRound()
+
+    if (activeRound && !this.#isRoundExpired(activeRound)) {
+      this.#currentRound = activeRound
+      console.log(`Resuming active round #${activeRound.id}`)
+    } else {
+      await this.#startNewRound()
+    }
+  }
+
+  /**
+   * Schedule periodic checks for round end
+   */
+  #scheduleRoundCheck () {
+    this.#checkRoundIntervalId = setInterval(async () => {
+      if (!this.#currentRound) return
+
+      const now = new Date()
+      if (new Date(this.#currentRound.end_time) <= now) {
+        try {
+          await this.#startNewRound()
+        } catch (error) {
+          console.error('Error handling round end:', error)
+        }
+      }
+    }, this.#config.checkRoundIntervalMs)
+  }
+
+  /**
+   * Start a new round
+   */
+  async #startNewRound () {
+    const previousRound = await this.#getActiveRound()
+    this.#currentRound = await this.#createNewRound()
+    if (!this.#currentRound) {
+      throw new Error('Failed to start a new round')
+    }
+
+    if (this.#taskingService) {
+      await this.#taskingService.generateTasksForRound(this.#currentRound.id)
+    }
+
+    if (previousRound) {
+      await this.#changeRoundActive(previousRound.id, false)
+    }
+
+    await this.#changeRoundActive(this.#currentRound.id, true)
+  }
+
+  /**
+   * Get the current active round from the database
+   */
+  async #getActiveRound () {
+    try {
+      const { rows } = await this.#db.query(`
+        SELECT * FROM checker_rounds 
+        WHERE active = true
+        ORDER BY start_time DESC
+        LIMIT 1
+      `)
+      return rows[0] || null
+    } catch (error) {
+      console.error('Error getting active round:', error)
```

There's a difference between finding no row and an error looking for rows. The 2nd should be bubbled up to the caller.

---

This function should throw error now instead of returning `null`.